File: 05-composicion-estaticos/java/Curso.java
/*
 * Curso.java - COMPOSICION
 *
 * Un Curso TIENE alumnos. No "es un" alumno: por eso NO hay herencia
 * aqui, sino un atributo que guarda otros objetos.
 *
 *   herencia    ->  ES UN    (Auto es un Vehiculo)
 *   composicion ->  TIENE UN (Curso tiene Alumnos)
 *
 * Esa pregunta, hecha en voz alta, resuelve el 90% de las dudas de diseno.
 */
import java.util.ArrayList;

public class Curso {

    private String nombre;
    private String sala;

    // COMPOSICION: este atributo guarda OTROS OBJETOS
    private ArrayList<Alumno> alumnos;

    public Curso(String nombre, String sala) {
        this.nombre = nombre;
        this.sala = sala;
        this.alumnos = new ArrayList<>();   // nace vacia, no null
    }

    public void inscribir(Alumno alumno) {
        this.alumnos.add(alumno);
        System.out.println(alumno.getNombre() + " inscrito en " + this.nombre);
    }

    public int cantidadAlumnos() {
        return this.alumnos.size();
    }

    public double promedio() {
        if (alumnos.isEmpty()) return 0;
        double suma = 0;
        for (Alumno a : alumnos) {
            suma += a.getNota();
        }
        return suma / alumnos.size();
    }

    public void listar() {
        System.out.println("Curso: " + nombre + " (" + sala + ")");
        for (Alumno a : alumnos) {
            System.out.printf("   - %-10s %.1f%n", a.getNombre(), a.getNota());
        }
    }

    public void mostrarAprobados(double minimo) {
        for (Alumno a : alumnos) {
            if (a.getNota() >= minimo) {
                System.out.println("   APROBADO: " + a.getNombre());
            } else {
                System.out.println("   reprobado: " + a.getNombre());
            }
        }
    }
}

File: 05-composicion-estaticos/python/curso.py
"""
curso.py - COMPOSICION

Un Curso TIENE alumnos. No "es un" alumno: por eso no hay herencia,
sino un atributo que guarda otros objetos.

CUIDADO con el error mas comun de quien viene de Java:
    class Curso:
        alumnos = []        # <- MAL: atributo de CLASE, se comparte
                            #    entre TODOS los cursos del programa
La lista tiene que crearse DENTRO de __init__, como esta abajo.
"""


class Curso:

    def __init__(self, nombre, sala):
        self._nombre = nombre
        self._sala = sala
        # COMPOSICION: este atributo guarda OTROS OBJETOS
        self._alumnos = []          # nace vacia, no None

    def inscribir(self, alumno):
        self._alumnos.append(alumno)
        print(f"{alumno.get_nombre()} inscrito en {self._nombre}")

    def cantidad_alumnos(self):
        return len(self._alumnos)

    def promedio(self):
        if not self._alumnos:
            return 0
        return sum(a.get_nota() for a in self._alumnos) / len(self._alumnos)

    def listar(self):
        print(f"Curso: {self._nombre} ({self._sala})")
        for a in self._alumnos:
            print(f"   - {a.get_nombre():<10} {a.get_nota():.1f}")

    def mostrar_aprobados(self, minimo):
        for a in self._alumnos:
            if a.get_nota() >= minimo:
                print(f"   APROBADO: {a.get_nombre()}")
            else:
                print(f"   reprobado: {a.get_nombre()}")
